applications/FluidDynamicsApplication/custom_utilities/mass_conservation_utility.cpp

```diff
@@ -30,131 +29,211 @@ namespace Kratos
+    Settings.ValidateAndAssignDefaults(default_parameters);
+    mpTimeVariable = &KratosComponents<Variable<double>>::Get(Settings["time_variable"].GetString());
+    mCorrectBackwards = Settings["correct_backwards"].GetBool();
+    mEchoLevel = Settings["echo_level"].GetInt();
+    mMinDt = Settings["dt_lower_limit"].GetDouble();
+    mMaxDt = Settings["dt_upper_limit"].GetDouble();
+    FindNodalHProcess<FindNodalHSettings::SaveAsNonHistoricalVariable> nodal_h_calculator(mrModelPart);
+    nodal_h_calculator.Execute();
 }
 
 
+
 /// Initialization function to find the initial volumes and print first lines in the log-file
-std::string MassConservationCheckProcess::Initialize(){
+std::string MassConservationUtility::Initialize(){
 
     double pos_vol = 0.0;
     double neg_vol = 0.0;
     double inter_area = 0.0;
-    const auto& r_comm = mrModelPart.GetCommunicator().GetDataCommunicator();
+    mrModelPart.GetCommunicator().SynchronizeVariable(DISTANCE);
+    mrModelPart.GetCommunicator().SynchronizeVariable(VELOCITY);
+    mrModelPart.GetCommunicator().SynchronizeVariable(MESH_VELOCITY);
 
     ComputeVolumesAndInterface( pos_vol, neg_vol, inter_area );
 
-    this->mInitialPositiveVolume = r_comm.SumAll(pos_vol);
-    this->mInitialNegativeVolume = r_comm.SumAll(neg_vol);
+    this->mInitialPositiveVolume = pos_vol;
+    this->mInitialNegativeVolume = neg_vol;
     this->mTheoreticalNegativeVolume = neg_vol;
-    inter_area = r_comm.SumAll(inter_area);
 
     std::string output_line =   "------ Initial values ----------------- \n";
     output_line +=              "  positive volume (air)   = " + std::to_string(this->mInitialPositiveVolume) + "\n";
     output_line +=              "  negative volume (water) = " + std::to_string(this->mInitialNegativeVolume) + "\n";
     output_line +=              "  fluid interface (area)  = " + std::to_string(inter_area) + "\n\n";
     output_line +=              "------ Time step values --------------- \n";
     output_line +=              "sim_time \t\twater_vol \t\tair_vol \t\twater_err \t\tnet_inf \t\t";
-    output_line +=              "net_outf \t\tint_area \t\tcorr_shift \n";
+    output_line +=              "net_outf \t\tint_area \n";
 
     return output_line;
 }
 
 
-std::string MassConservationCheckProcess::ExecuteInTimeStep(){
+
+std::string MassConservationUtility::ComputeBalancedVolume(){
 
     // performing all necessary calculations
     double pos_vol = 0.0;
     double neg_vol = 0.0;
-    double inter_area = 0;
+    double inter_area = 0.0;
+    mrModelPart.GetCommunicator().SynchronizeVariable(DISTANCE);
+    mrModelPart.GetCommunicator().SynchronizeVariable(VELOCITY);
+    mrModelPart.GetCommunicator().SynchronizeVariable(MESH_VELOCITY);
+
     ComputeVolumesAndInterface( pos_vol, neg_vol, inter_area );
     double net_inflow_inlet = ComputeFlowOverBoundary(INLET);
     double net_inflow_outlet = ComputeFlowOverBoundary(OUTLET);
```

Comment:
Maybe it is a good idea that in the `Initialize` function we check that there exists entities that are flagged with `INLET` and `OUTLET` flags.
Although cases with neither inlet nor outlet are valid (e.g. sloshing thank)  I think we should at least throw a warning. This would make non-experienced users life easier.